omniduct/filesystems/s3.py

```diff
@@ -0,0 +1,154 @@
+        try:
+            self._client.get_object(Bucket=self.bucket, Key=path or '')
+            return True
+        except:
+            return False
+
+    def __dir_paginator(self, path):
+        path = path or ''
+        if path.endswith('/'):
+            path = path[:-1]
+        paginator = self._client.get_paginator('list_objects')
+        iterator = paginator.paginate(
+            Bucket=self.bucket,
+            Prefix=path + ('/' if path else ''),
+            Delimiter='/',
+            PaginationConfig={'PageSize': 500}
+        )
+        return iterator
+
+    def _listdir(self, path):
+        iterator = self.__dir_paginator(path)
+
+        paths = []
+        for response_data in iterator:
+            for prefix in response_data.get('CommonPrefixes', []):
+                paths.append(prefix['Prefix'][:-1])  # Remove trailing slash
+            for prefix in response_data.get('Contents', []):
+                paths.append(prefix['Key'])
+
+        return paths
+
+    def _showdir(self, path):
+        iterator = self.__dir_paginator(path)
+
+        paths = []
+        for response_data in iterator:
+            for prefix in response_data.get('CommonPrefixes', []):
+                paths.append({
+                    'type': 'dir',
+                    'name': prefix['Prefix'][:-1].split('/')[-1]  # Remove trailing slash
+                })
+            for prefix in response_data.get('Contents', []):
+                paths.append({
+                    'type': 'file',
+                    'name': prefix['Key'].split('/')[-1],
+                    'last_modified': prefix['LastModified'],
+                    'owner': prefix['Owner']['DisplayName'],
+                    'size': prefix['Size']
+                })
+
+        df = pd.DataFrame(paths)
+        col_order = [col for col in ('type', 'name', 'size', 'owner', 'last_modified') if col in df.columns]
+        return df[col_order]
+
+    def _find(self, pattern, path_prefix, files, dirs):
+        if dirs:
+            logger.warning('S3Client does not support finding directories.')
+        if not files:
+            return
+
+        path_prefix = path_prefix if path_prefix is not None else ''
+        pattern = re.compile(pattern) if pattern is not None else None
+
+        b = self._resource.Bucket(self.bucket)
+        keys = b.objects.filter(Prefix=path_prefix)
+        for k in keys:
+            if pattern is None or pattern.match(k.key[len(path_prefix):]):
+                yield k.key
+
+    # File opening
+
+    def _file_read_(self, path, size=-1, offset=0, binary=False):
+        assert self.isfile(path), "File `{}` does not exist.".format(path)
+        obj = self._resource.Object(self.bucket, path)
+        body = obj.get()['Body'].read()
+
+        if not binary:
+            body = body.decode()
+
+        return body
+
+    def _file_append_(self, path, s, binary):
+        raise NotImplementedError
+
+    def _file_write_(self, path, s, binary):
```

Comment:
question is this because we want this to be read only or because boto doesn't support it?

Perhaps the exception should be written to express which circumstance we're in.

Reply:
Boto supports it, but I have yet to test it on an S3 filesystem I have access to. Will add it in the future. Fixed.